modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -130,16 +160,52 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
   )(implicit
     F: Monad[F],
     C: Clock[F]
-  ): F[Option[SchemaLookup]] =
+  ): F[Option[(SchemaLookup, Int)]] =
     cache match {
-      case Some(c) => c.getSchema(schemaKey)
+      case Some(c) => c.getTimestampedSchema(schemaKey)
       case None    => Monad[F].pure(None)
     }
 }
 
 /** Companion object. Lets us create a Resolver from a Json */
 object Resolver {
 
+  /** The result of doing a lookup with the resolver, carrying information on whether the cache was used */
+  sealed trait ResolverResult[+A] { self =>
+
+    /**
+     * Convert the result to an `Either`
+     *
+     *  @return either the failure (as left) or a tuple of the looked-up item and a timestamp of
+     *  when the item was added to the cache (as right)
+     */
+    def toEither: Either[ResolutionError, (A, Option[Int])] =
+      self match {
+        case Cached(value, t)   => Right((value, Some(t)))
+        case NotCached(value)   => Right((value, None))
+        case LookupError(value) => Left(value)
+      }
+  }
+
+  /**
+   * The result of a lookup when the resolver is configured to use a cache
+   *
+   *  The timestamped value is helpful when the client code needs to perform an expensive
+   *  calculation derived from the looked-up value. If the timestamp has not changed since a
+   *  previous call, then the value is guaranteed to be the same as before, and the client code
+   *  does not need to re-run the expensive calculation.
+   *
+   *  @param value the looked-up value
+   *  @param timestamp epoch time in seconds of when the value was last cached by the resolver
+   */
+  case class Cached[A](value: A, timestamp: Int) extends ResolverResult[A]
```

Comment:
Can we add key here as well? And then pass whole thing to validator, without converting to `Option`?